Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/test/core/TestCertificateRepository.kt

```diff
@@ -133,7 +141,38 @@ class TestCertificateRepository @Inject constructor(
             mutate { this[container.identifier] = container }
         }
 
-        return newData.values.single { it.registrationToken == test.registrationToken }
+        return newData.values.single {
+            it.data is RetrievedTestCertificate && it.data.registrationToken == test.registrationToken
+        }
+    }
+
+    suspend fun registerTestCertificate(
+        qrCode: TestCertificateQRCode
+    ): TestCertificateContainer {
+        Timber.tag(TAG).v("registerTestCertificate(qrCode=%s)", qrCode)
+
+        val updatedData = internalData.updateBlocking {
+            val identifier = UUID.randomUUID().toString()
+
+            val nowUtc = timeStamper.nowUTC
+
+            val data = GenericTestCertificateData(
+                identifier = identifier,
+                registeredAt = nowUtc,
+                certificateReceivedAt = nowUtc,
+                testCertificateQrCode = qrCode.qrCode
+            )
+            val container = TestCertificateContainer(
+                data = data,
+                qrCodeExtractor = qrCodeExtractor,
+            )
+            Timber.tag(TAG).d("Adding test certificate entry: %s", container)
+            mutate { this[container.identifier] = container }
+        }
+
+        // We just registered it, it MUST be available.
+        return updatedData.values
+            .single { it.certificateId == qrCode.uniqueCertificateIdentifier }
     }
```

Comment:
Maybe here?

Reply:
Ah, good catch. I'm not checking for existing certificates for scanned certs, only when retrieved.
Will fix.

Reply:
Fixed.